### PR TITLE
Use full qualified name for PythonOp export

### PR DESCRIPTION
**Pull request by pengwa**
### Use full qualified name for PythonOp export

Originally, when there are duplicate named torch.autograd.Function in different module, for example:

`a.b.c.Gelu` v.s. `d.e.func.<locals>.Gelu`

We by default will throw exception to let user be aware we cannot distinguish the two Gelu because during model export, we did not module path. The workaround is we introduced `ORTMODULE_SKIPPED_AUTOGRAD_FUNCTIONS` to ignore those duplicated named Gelu that is not used by model run. This has limitations obviously for example if two Gelus are both used in training. 

This PR finds a way to construct a full qualified name.

`def _export_pt_1_10(g, n, *args, **kwargs):`

1. in exporter function, kwargs contains `name` and `module`, in the above example:
   `a.b.c.Gelu`  --> name: `Gelu`, module: `a.b.c`
   `d.e.func.<locals>.Gelu` --> name: `Gelu`, module: `d.e`
   
 
    Using name and module is not enough to get a full qualified name, for the second case, where `d.e` is the module path, then there is a function called `func`, in this function, there is a local auto.grad.Function named `Gelu`. (Many of our UT looks like this). We can only get `d.e.Gelu`, but this is not the correct full qual name.

    The reason for this: `kwargs[name]` or `n.name` only return the class's name, not the class's full qual name. (be noted kwargs[module]` is correct).

2. `n`  is torch.Node, we can access `pyobj` to get the torch.autograd.Function's apply method instance, then use `._self` to get the torch.autograd.Function class. Then we can get the `module` and `class`'s ful qual name, added together, we get the full qual name.

With the above change, we don't need use `kwargs[name]` and `kwargs[module]` , and don't need check naming conflicting or `ORTMODULE_SKIPPED_AUTOGRAD_FUNCTIONS` env var any more.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_utils.py [hunk 1]
```diff
@@ -419,6 +419,13 @@ def check_function_has_param(function: Callable, param_name: str) -> bool:
     return param_name in inspect.signature(function).parameters
 
 
+def get_fully_qualified_class_name(klass):
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/ortmodule/_utils.py:422]:
```suggestion
def get_fully_qualified_class_name(klass) -> str:
```

I would add types to all newly defined functions

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py [hunk 2]
```diff
@@ -204,68 +232,27 @@ def _export_pt_1_10(g, n, *args, **kwargs):
...
-            break
-
-    if is_pythonop_needed:
-        warnings.warn(
-            "Detected autograd functions usage in current model, the run will fail \
-                      without enabling '_enable_custom_autograd_function'. Please enable it with: \
-                      'module._execution_manager(is_training_mode)._enable_custom_autograd_function = True'",
-            UserWarning,
-        )
-
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:221]: seems unrelated to the PR goal. Is that intentional?

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py [hunk 3]
```diff
@@ -204,68 +232,27 @@ def _export_pt_1_10(g, n, *args, **kwargs):
...
-        full_qualified_name = _full_name(kclass)
-        if full_qualified_name in skipped_autograd_function_list:
-            continue
-        # Collect mapping of class names to full qualified class names.
-        if kclass.__name__ not in registered_name_mappings:
-            registered_name_mappings[kclass.__name__] = []
-        registered_name_mappings[kclass.__name__].append(full_qualified_name)
-
-        # Register function with class names.
-        register_torch_autograd_function(kclass.__name__, kclass)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:240]: seems unrelated to the PR goal. Is that intentional?

File: orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc [hunk 4]
```diff
@@ -309,7 +309,8 @@ void IterateSubgraphFromNode(Graph& graph,
         continue;
       }
       auto func_name = static_cast<std::string>(cur->GetAttributes().at("name").s());
-      if (func_name == "_InspectActivation" || func_name == "_IncrementStep") {
+      if (func_name == "onnxruntime.training.utils.hooks._subscriber_manager._InspectActivation" ||
+          func_name == "onnxruntime.training.utils.hooks._subscriber_manager._IncrementStep") {
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc:313]: `"onnxruntime.training.utils.hooks._subscriber_manager._InspectActivation"` and `"onnxruntime.training.utils.hooks._subscriber_manager._IncrementStep"` are used over and over. maybe a constant to prevent typos that would cause bugs?

Comment (baijumeswani) [inline orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc:313]:
```suggestion
      if (func_name in ["onnxruntime.training.utils.hooks._subscriber_manager._InspectActivation", "onnxruntime.training.utils.hooks._subscriber_manager._IncrementStep"]) {
```

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py (hunk 2 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:221]: Oh I fortgot to mention, this is kind of dead code, that won't be trigger. If enable_custom_autograd_function, then this won't be run; otherwise, export will fail to handle autograd.Function because we did not register custom export function; So anyway, this won't be triggered. :)

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py (hunk 3 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:240]: Oh, this is by intention, I mentioned it a bit in the last paragram "With the above change, we don't need use kwargs[name] and kwargs[module] , and don't need check naming conflicting or ORTMODULE_SKIPPED_AUTOGRAD_FUNCTIONS env var any more.". 

Since we can get the autograd.Function classes in the exporter function, are using full qual name now, we can just do the registration there, don't need do this after exporting. There is a side good effect: for those not used torch.autograd.FUnction.__subclass__(), we don't register them any more.

Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:240]: BTW, register_torch_autograd_function tolerant repeated registry, it respect the first time registration behavior. ;)

File: orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc (hunk 4 above)
Comment (pengwa) [reply orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc]: The concern makes senses. I defined constant std::string for now locally as a temporary fix, the longer term fix is: we had a work to allow defining customized shape inferencing from Python binding APIs, then we won't have such named based check any more later.

Comment (pengwa) [reply orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc]: I believe you thought it is orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.py. :)

**Review by pengwa — COMMENTED** [review]
Thanks a lot for the quick actions! I refined based on suggestions. :)

**Review by askhade — APPROVED** [review]
LGTM

**Review by baijumeswani — APPROVED** [review]

**Comment by pengwa** [comment]
Thanks a lot, @askhade @baijumeswani  @justinchuby  @thiagocrepaldi!!!
